### PR TITLE
MNT: Force N to be int in _scipy_kraft_burrows_nousek

### LINKED ISSUE #10832: `stats.funcs._scipy_kraft_burrows_nousek` raises DeprecationWarning under Python 3.9
### Description

As of Python 3.9.0 `test_scipy_poisson_limit` fails with a warning
```
>       assert_allclose(funcs._scipy_kraft_burrows_nousek(5., 2.5, .99),
                        (0, 10.67), rtol=1e-3)
...
>       factorial_N = float(math.factorial(N))
E       DeprecationWarning: Using factorial() with floats is deprecated
```
due to the function calling `math.factorial` on a `float` input:
https://github.com/astropy/astropy/blob/89580eee52dbe06e54fd51abe69ba0b0ecefc33a/astropy/stats/funcs.py#L1120

### Expected behavior

The function actually documents the input as `N : int or np.int32/np.int64`, but should check this or convert using `round`.

### Actual behavior
See above.

### Steps to Reproduce
```python
import astropy
astropy.test(package='stats', args='-vk scipy_poisson')
```

### System Details
macOS-10.14.6-x86_64-i386-64bit
Python 3.9.0 (default, Oct 10 2020, 10:27:03) 
[Clang 11.0.0 (clang-1100.0.33.17)]
Numpy 1.19.2
astropy 4.1rc2
astropy 4.2.dev863+ge114134a2
Scipy 1.5.2
Matplotlib 3.3.2

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to address deprecation warning when `N` is not an integer in Python 3.9. Not sure why our cron job didn't catch it, but regardless, I don't see how it can hurt to force `N` to be `int` anyway.

I originally thought about throwing an exception but from the tests, it seems like the function is supposed to accept `N` as `float` even when the docstring says `int`.  🤷 

Fixes #10832

### CODE REVIEW

**Review by dhomeier — CHANGES_REQUESTED** [review]
I had also been thinking about raising an exception or another warning (and intercept/ignore that in the tests), but the existing test indeed suggests to just let it pass.
My example on the roundoff errors however shows that this can in fact produce different results if by mistake an actual non-integer value is passed:
`math.factorial(61 / 7 * 7)` already raises a `ValueError`, not just a warning, as the input is non integer-valued. With this fix such inputs would now be silently accepted.
So if we want to be really careful, we should perhaps check something like
```
    if not isinstance(N, int):
        N_in = N
        N = round(N)
        if abs(N - N_in) > 1.e-3:
            warnings.warn(f"input N to `_scipy_kraft_burrows_nousek` was rounded from {N_in} to {N}!",
                          AstropyWarning)
```

**Comment by pllim** [comment]
I guess the question is just how careful we want to be when API clearly states `int` only but somehow the function silently accepts float. 🤷

File: astropy/stats/funcs.py
Comment (pllim) [reply astropy/stats/funcs.py:1107]: Thanks! I'll wait a bit for others to chime in before applying the suggestions.

**Comment by dhomeier** [comment]
Yes, the docstring is clear enough – my concern was just that this would now accept "more inaccurate" inputs, so at least wanted to put this up for discussion.  Could be a case for `typing_extensions`...

**Comment by pllim** [comment]
@larrybradley , any advice?

**Comment by pllim** [comment]
Doesn't look like this will get resolved in time, so moving milestone.

**Comment by dhomeier** [comment]
I'm OK without additional warnings, since it is not a public function; also there are two implementations, and `_mpmath_kraft_burrows_nousek` in contrast explicitly casts its `N` to `float`.
But I'd still like at least to hear a second opinion before using `int(N)` instead of `round(N)`, as this could in some (rare) cases change results.

**Comment by larrybradley** [comment]
If we are going to allow `N` to be float, then `round(N)` makes more sense than `int(N)`.

I haven't looked at these functions in detail -- is `N` ever float?

**Comment by dhomeier** [comment]
> I haven't looked at these functions in detail -- is `N` ever float?

It is a count number, thus should always be integer-valued, but as discussed in https://github.com/astropy/astropy/pull/10900/files/094254c5513ae7d28c28ceb9b0798f42f79f9793#r512177609 there is already fp-math done on it in other places, and in `_mpmath_kraft_burrows_nousek` throughout.

Since this is milestoned to 4.0.4, using `round(N)` here is probably the better solution rather than leaving the fix to #10900.

**Comment by larrybradley** [comment]
@dhomeier Sounds good.  Let's go with `round(N)`.

**Comment by pllim** [comment]
Okay, I applied @dhomeier 's suggestion above. Thanks! Is there anything else need to be done here? As far as I can understand https://github.com/astropy/astropy/pull/10900/files#r512177609 , this is all we need?

**Comment by hamogu** [comment]
Let me way in as the original author of that code. I've tried to dig back through issues, but some of this is from memory. Scientifically, there is no need for this function to accept floating point values for N, in fact is should not. I'm not certain why I did that initially, it may be because back in Python 2, "4 / 3" would give you an integer. If you "blame" on the tests, you'll see that the tests with the floats are original, while the int tests with ints are a recent addition. In that sense, the tests don't intentionally test that floats work, they just happen to be the older ones that are still around.

I suggest we take this opportunity to not accept float any longer. I would argue that accepting floats was a bug in the first place and thus can be fixed without a deprecation period.

**Comment by hamogu** [comment]
Sorry, I'm late to the party, I don't browse open PRS ands issues as often as you guys do...
To be clear, I suggest to check that N is an int type (ind, np.int32 ,...) in the "master" function around this line: https://github.com/astropy/astropy/blob/master/astropy/stats/funcs.py#L747

Where needed (e.g. here https://github.com/astropy/astropy/blob/master/astropy/stats/funcs.py#L1106) n is then cast to float again. That's because floats can store much larger values (albeit at a loss of precision) compared to ints.

Then there is this: https://github.com/astropy/astropy/blob/master/astropy/stats/funcs.py#L1180
That's not because N *should* be a float, but just because mpmath (at least in the version that I have) cannot create the numeric format that we need later (mpf) directly from int. mpmath is a floating point library, so we'll have to make it float, just so we can call mpmath function on it, but that's an iplementation detail that users should not know about, they can just put in an int for N.

**Comment by dhomeier** [comment]
Thanks for the confirmation; so the `TypeError` should be raised in the publicly visible `poisson_conf_interval`, right?

Since the test in question is directly calling the _scipy function, that would either need a special case for Python 3.9+ (or ignore the warning, but in 3.10 or 3.11 it should become an error); or `_scipy_kraft_burrows_nousek` should just keep the `round(N)` as additional safeguard.

**Comment by hamogu** [comment]
I agree to the `TypeError`. I don't think any additional safeguard is needed in the `_scipy_kraft_...` function. If you are using an underscore function, you know you are on your own. If it's used with the wrong type of argument (a float instead of an int) it *is* going to fail with an error or a warning on those Python versions that don't support the float argument; only that error comes from the math library and not from astropy. I think that's perfectly acceptable. No need for us to write, test, and maintain an additional check.

**Comment by dhomeier** [comment]
> If you are using an underscore function, you know you are on your own. If it's used with the wrong type of argument (a float instead of an int) it _is_ going to fail with an error or a warning on those Python versions that don't support the float argument; only that error comes from the math library and not from astropy. I think that's perfectly acceptable. No need for us to write, test, and maintain an additional check.

Well, the test is already there; as it behaves differently depending on the Python version, this triggered the original issue, and will require maintenance one way or another if it's not caught in the function itself. But handling this with a version-dependent `pytest.warns` or `pytest.raises` might indeed be the better way to deal with it.

**Comment by pllim** [comment]
How does this look now? With the suggested changes, it can no longer be backported.

When we must test 4.0.x or 4.1.x, I think we can just modify the tests in those branches to ignore the warning. But we'll cross that bridge when we can add Python 3.9 to the matrix (currently lacking Numpy and Scipy wheels).

File: astropy/stats/funcs.py
Comment (dhomeier) [inline astropy/stats/funcs.py:753]:
```suggestion
        if not issubclass(np.array(n).dtype.type, np.integer):
            raise TypeError('Number of counts must be integer.')
```
Could be shortened a bit, I think, or would this be missing some case?

**Review by dhomeier — APPROVED** [review]
Looks good, and the tests much tidier!
Approving with the suggestion on the type check or without.

But also spotted some more tests to be fixed: there are a bunch of calls with `np.zeros` and `np.ones` from
https://github.com/astropy/astropy/blob/e1a38f38ff78430fa3ed9436c69de13cfcd58995/astropy/stats/tests/test_funcs.py#L457-L459

and following!

File: astropy/stats/funcs.py
Comment (pllim) [reply astropy/stats/funcs.py:753]: Any suggested new verbiage I should use?

Comment (pllim) [reply astropy/stats/funcs.py:753]: Oh, you mean the `if` logic, not the error message. Sorry for the misunderstanding.

Please let me know how I can make the check more robust. I see that it handles scalar and ndarray, so I thought this logic would be enough, but I am open to suggestions.

Comment (dhomeier) [reply astropy/stats/funcs.py:753]: No, I think it was OK, just seemed more code than necessary, but I was not sure if _my suggestion_ fully replaced it. But I think it is sufficient, too.

Comment (pllim) [reply astropy/stats/funcs.py:753]: Ah, sorry, looks like I missed that you provided a suggestion too. 😬  I guess performance hit of converting a scalar to array is negligible, all things considered, so it boils down to preference? @larrybradley and @hamogu , what say you?

**Review by dhomeier — COMMENTED** [review]
On more thing...

I asked to fix the tests using np.ones(i) and the like as input, since I thought the exception would be always raised right after entering the function, so I was a bit surprised when all of those still passed. But I am seeing now the @hamogu indeed suggested to add it just in the `kraft-burrows-nousek` interval case.
Per the function docstring and conceptually it is _always_ an integer, although for many of the other interval types it is immediately used as a float. Wondering if it would be less confusing to consistently reject non-integers.

**Comment by hamogu** [comment]
The KBN is the case I know about and the case that's discussed in this PR. We should ask @larrybradley what he thinks about the more general case. To be clear, I'm not opposed to have ll of them "do the right thing" and expect integer counts, but I also know that for convenience it's often useful to accept floats for ints, too. In KBN that can cause issues (as we saw) so we don't want that, but in many other cases, it won't (e.g. 1/sqrt(N), will work just find with floats and won't really loose any precision compared to ints).

So, I suggest to merge it like this just for the KBN case.

**Review by hamogu — APPROVED** [review]

**Comment by larrybradley** [comment]
@hamogu That sounds fine with me.

**Review by larrybradley — APPROVED** [review]

**Comment by pllim** [comment]
Thanks, everyone!

If some stop-gap fix is needed for astropy 4.0.x or 4.2 in the future when #10832 crops up again when we add Python 3.9 to the matrix, I think we can simple accept the warning for those. This PR is going into 4.3 unless @bsipocz thinks it is not too late for 4.2.

**Comment by bsipocz** [comment]
This should still be fine for 4.2, as that will certainly support 3.9 (not sure for 4.0), and I haven't yet done the RC. Thanks for the heads up.

**Comment by pllim** [comment]
@bsipocz , if you need me to open a follow up PR to move the change log entry, please let me know. Thanks!

**Comment by bsipocz** [comment]
no worries, I'll fix the changelog while backporting, and will propagate back everything once the RC is done.

**Pull request merged**
